content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+
+
+### Register Promotion
+
+Whenever a shared memory location is read and modified, it must first **promote** its value in the memory hierarchy: that is, it must move its value into a local register. 
+After operating on this promoted value, it must then **demote** it by moving the computed value back into the shared memory location. 
+Such promotion/demotion operations are costly, and especially so in a loop.
+
+Register promotion is an optimization that aims to minimize the cost of promotion/demotion operations in a loop by maximizing data locality at the register level.
+Put another way, it aims to ‘factor out’ promotions in a loop.
+This optimization transforms a loop by first speculatively promoting the value from a shared variable accessed in the loop into a new, local variable, before the loop. 
+Next, all usages of the shared variable are replaced by this new promoted variable within the loop. 
+Finally, the promoted variable is demoted after the loop.
+If applied to lock-synchronized code, the promotion/demotion operations added by register promotion will lie outside the critical section, therefore creating a data race.
+
+Importantly, register promotion must take care to insert a pair of promotion/demotion operations around opaque function calls within the loop body.
+This is because such functions are assumed to refer to all shared variables, so a demotion is required before the call, followed by a promotion.
+Register promotion is therefore not beneficial in general, since it may increase the number of promotions and demotions. 
+A heuristic must be employed to determine whether its usage improves performance.
+
+
+To illustrate how register promotion can create a data race, consider the following example, where a shared variable `x` is modified:
+```c
+for (...) {
+  ...
+  if (mt) pthread_mutex_lock(...);
+  x = ... x...
+  if (mt) pthread_mutex_unlock(...);
+}
+```
+
+Suppose `mt` is only true when multiple threads have been created.
+Then this program lacks data races, since reads and writes to `x` are logically synchronized.
+
+Further suppose that register promotion is determined to be a beneficial optimization, perhaps because the conditional is observed to rarely be taken.
+Since the Pthreads library calls are simply regarded as opaque function calls, the optimization will produce the following:
+
+```c
+r = x; 
+for (...) {
+  ...
+  if (mt) {
+    x = r; pthread_mutex_lock(...); r = x;
+  }
+  r = ... r...
+  if (mt) {
+    x = r; pthread_mutex_unlock(...); r = x;
+  }
+}
+x = r; 
+```
+
+In this case, register promotion introduces unsynchronized reads and writes to `x`, creating a data race.
```

Comment:
Clarify where the races are—in a multithreaded execution, which access races with which other access? And again, remind the reader why this is bad: it took race-free code (with well-defined semantics) and turned it into bad code with undefined semantics.